smart-contract-extensions/contracts/KeyPurchaser.sol

```diff
@@ -0,0 +1,235 @@
+import '@openzeppelin/contracts/math/SafeMath.sol';
+import 'unlock-abi-7/IPublicLockV7.sol';
+
+/**
+ * @notice Purchase a key priced in any ERC-20 token - either once or as a regular subscription.
+ * This allows the user to purchase or subscribe to a key with 1 tx (`approve`)
+ * or if the token supports it, with 1 signed message (`permit`).
+ *
+ * The user can remove the ERC-20 approval to cancel anytime.
+ *
+ * Risk: if the user transfers or cancels the key, they would naturally expect that also cancels
+ * the subscription but it does not. This should be handled by the frontend.
+ */
+contract KeyPurchaser is Initializable, Stoppable
+{
+  using Address for address payable;
+  using SafeERC20 for IERC20;
+  using SafeMath for uint;
+
+  // set on initialize and cannot change
+
+  /**
+   * @notice This is the lock for the content users are subscribing to.
+   */
+  IPublicLockV7 public lock;
+
+  /**
+   * @notice The most you will spend on a single key purchase.
+   * @dev This allows the lock owner to increase the key price overtime, up to a point, without
+   * breaking subscriptions.
+   */
+  uint public maxPurchasePrice;
+
+  /**
+   * @notice How close to the end of a subscription someone may renew the purchase.
+   * @dev To ensure that the subscription never expires, we allow this amount of time (maybe 1 day)
+   * before it expires for the renewal to happen.
+   */
+  uint public renewWindow;
+
+  /**
+   * @notice The earliest a renewal may be purchased after the previous purchase was done.
+   * @dev This typically would not apply, but helps to avoid potential abuse or confusion
+   * around cancel and transfer scenarios.
+   */
+  uint public renewMinFrequency;
+
+  /**
+   * @notice The amount of tokens rewarded from the end user to the msg.sender for enabling this feature.
+   * This is paid with each renewal of the key.
+   */
+  uint public msgSenderReward;
+
+  // admin can change these anytime
+
+  /**
+   * @notice Metadata for these subscription terms which may be displayed on the frontend.
+   */
+  string public name;
+
+  /**
+   * @notice Metadata for these subscription terms which can be used to supress the offering
+   * from the frontend.
+   */
+  bool internal hidden;
+
+  // store minimal history
+
+  /**
+   * @notice Tracks when the key was last purchased for a user's subscription.
+   * @dev This is used to enforce renewWindow and renewMinFrequency.
+   */
+  mapping(address => uint) public timestampOfLastPurchase;
+
+  /**
+   * @notice Called once to set terms that cannot change later on.
+   * @dev We are using initialize instead of a constructor so that this
+   * contract may be deployed with a minimal proxy.
+   */
+  function initialize(
+    IPublicLockV7 _lock,
+    address _admin,
+    uint _maxPurchasePrice,
+    uint _renewWindow,
+    uint _renewMinFrequency,
+    uint _msgSenderReward
+  ) public
+    initializer()
+  {
+    _initializeAdminRole(_admin);
```

Comment:
On second thought, I'll remove the admin role and use the lock's lockManagers instead.